### ISSUE TITLE
ServerOnlyStreamInterface::sendAndClose does not raise exceptions

### ISSUE THREAD

**Issue by milgner** [connectrpc/connect-kotlin#262]
I'm trying to establish a stream connection to a remote and deal with connection issues.
My code to establish the response channel currently looks like this:

```
val receiveData = sequencerClient.receiveData() // it's a ServerOnlyStreamInterface
val receiveDataRequest = receiveDataWithPreloadSelector(preloadSelector) // just the request input
receiveData.sendAndClose(receiveDataRequest) // always succeeds, even if the remote is not available
receiveData.responseChannel()
```

if the remote service is not available, I would have expected the invocation of `sendAndClose` to raise an error. However, it looks like `sendAndClose` doesn't actually send anything. Only when invoking `receiveCatching` on the response channel do I get an exception that the remote is not available.

However, I really want to have different error handling behaviour for establishing the initial connection and dealing with subsequent issues: if it cannot establish the connection, it's not so bad - but if something breaks while streaming data of an existing connection, it needs to handle it more urgently.

Of course I can introduce a new state variable that indicates whether something was already read etc. But I find it quite irritating that a method called `sendAndClose` doesn't raise an exception when it cannot send anything.

**Comment by jhump** [comment]
@milgner, I suspect you are running into a sharp edge of the API. If I am correct, the problem is that the `sendAndClose` method currently returns a `Result<Unit>`. This _intentionally_ never throws an exception as the `Result` type transforms exceptions into values.

So when you say it always succeeds, do you mean that it never throws an exception? Or is your real code actually examining the result value to see if it was successful? Unfortunately, you must do the latter: you must not ignore the return value (as in your example snippet above) but actually interrogate it.

This is something we will fix in the API before this repo reaches v1.0: to get rid of `Result<Unit>` from the API entirely and use exceptions instead. The `Result` type is really only valuable for functions that return an actual value that the caller wants to examine. So the net effect of using `Result<Unit>` is that failures are quiet and too easy to miss.

**Comment by milgner** [comment]
> So when you say it always succeeds, do you mean that it never throws an exception? Or is your real code actually examining the result value to see if it was successful? Unfortunately, you must do the latter: you must not ignore the return value (as in your example snippet above) but actually interrogate it.
>

Oh my... I somehow totally overlooked the fact that a `Result` is being returned here :see_no_evil: 
 
> This is something we will fix in the API before this repo reaches v1.0: to get rid of `Result<Unit>` from the API entirely and use exceptions instead. The `Result` type is really only valuable for functions that return an actual value that the caller wants to examine. So the net effect of using `Result<Unit>` is that failures are quiet and too easy to miss.

In principle I'm a big fan of using `Result` and `Option` types and using `map` et al to transform them. But since other methods used to throw exceptions, I wasn't looking for a `Result` type at all... I'm just gonna change the code accordingly!

**Comment by jhump** [comment]
> In principle I'm a big fan of using `Result` and `Option` types and using map et al to transform them. But since other methods used to throw exceptions, I wasn't looking for a `Result` type at all...

That's fair: API consistency/coherence is something we will address as well.

FWIW, I do like the use of functional types like `Result` and `Option`, but they don't really make as much sense to combine with `Unit`: since that is really the _absence_ of a value, many users won't think to use the functional/monadic methods like `map`, etc. Also, I think block-based APIs are better for managing the lifetime of resources (and that is the direction we plan to take the stream-related methods), and they lend themselves to exception-based control flow.
